imageMaximumGCAge download /remove loop with CreateContainerConfigError state

aryasoni98 opened the pull request:
<!--  
Thanks for sending a pull request! Please read the guidelines:
https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution  
https://git.k8s.io/community/contributors/devel/development.md#development-guide  
-->

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

This PR fixes an issue where images used by containers in error states (such as `CreateContainerConfigError`) were repeatedly garbage collected and re-downloaded, causing unnecessary network traffic and resource usage.  
The image garbage collection logic now excludes containers that are not actually using their images (i.e., not in "running" or "created" state), preventing images referenced by error-state containers from being considered "in use."  
This avoids the download/remove loop and reduces unnecessary image pulls.

#### Which issue(s) this PR is related to:

Fixes #132403

#### Special notes for your reviewer:

- The fix introduces a helper function to determine if a container is actually using its image.
- Unit tests and image GC tests have been updated to reflect the new logic.
- Please review the test changes to ensure they align with the intended behavior.

#### Does this PR introduce a user-facing change?

```release-note
Kubelet: Fixes an issue where images used by containers in error states (e.g., CreateContainerConfigError) were repeatedly garbage collected and re-downloaded, causing high network traffic. Images are now only considered "in use" if the container is actually running or created.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
N/A
```

aryasoni98 committed:
imageMaximumGCAge download /remove loop with CreateContainerConfigError state

bart0sh commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -256,6 +280,16 @@ func (im *realImageGCManager) detectImages(ctx context.Context, detectTime time.
 		for _, container := range pod.Containers {
+			// Skip containers that are not actually using the image
+			// This includes containers in error states like CreateContainerConfigError
+			if !isContainerActuallyUsingImage(container) {
```
Does this mean that if image is not used by any container it will never be gc-ed?

aryasoni98 replied:
@bart0sh  No, that's not the case.

If an image is not used by any container, it is eligible for garbage collection according to the usual image GC policies (such as imageMaximumGCAge and disk pressure thresholds).

This change only ensures that images used by containers in error states (like CreateContainerConfigError) are not considered "in use" and can be garbage collected if not used elsewhere.

Let me know if you have any other questions!

bart0sh replied:
> This change only ensures that images used by containers in error states (like CreateContainerConfigError) are not considered "in use" and can be garbage collected if not used elsewhere.

How does this fix the case described in the PR description: `images used by containers in error states (such as CreateContainerConfigError) were repeatedly garbage collected and re-downloaded, causing unnecessary network traffic and resource usage.` ?
According to the change, images referenced by containers with `CreateContainerConfigError` will not be considered as `in use` and therefore will be gc-ed and re-downloaded, no?

bart0sh replied:
@aryasoni98 Please, address my above comment, thanks!

bart0sh commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -240,6 +240,30 @@ func (im *realImageGCManager) GetImageList() ([]container.Image, error) {
+// isContainerActuallyUsingImage determines if a container is actually using its image.
+// Containers in error states or not running are not considered to be using their image
+// for garbage collection purposes.
+func isContainerActuallyUsingImage(container *container.Container) bool {
+	// Only consider containers that are actually running or in a valid state
+	// that indicates they are using the image
+	switch container.State {
+	case "running":
+		return true
+	case "created":
+		// Created containers may be using the image but haven't started yet
+		return true
+	case "exited":
+		// Exited containers are not currently using the image
+		return false
+	case "unknown":
+		// Unknown state containers are not considered to be using the image
+		return false
+	default:
+		// For any other state, be conservative and assume they're not using the image
+		return false
+	}
+}
```
nit:
```suggestion
func isContainerActuallyUsingImage(c *container.Container) bool {
	// Only consider containers that are actually running or created
	// that indicates they are using the image
	return c.State == container.ContainerStateRunning || c.State == container.ContainerStateCreated	
}
```

If you prefer to use switch, please consider at least using constants from the container module instead of hardcoded strings.

aryasoni98 committed:
Fix: use container state constants in isContainerActuallyUsingImage to avoid hardcoded strings and shadowing issues

aryasoni98 commented:
Hi @bart0sh, just a gentle follow-up on the PR when you get a chance.

If everything looks good, would appreciate it if you could proceed with the merge.
Thanks again for your time and guidance!

aryasoni98 committed:
imageMaximumGCAge download /remove loop with CreateContainerConfigError state

haircommander commented:
@aryasoni98 thanks for opening this! I don't love how this duplicates logic between multiple places and also makes the GC logic quite a bit more complicated. I left a suggestion in https://github.com/kubernetes/kubernetes/issues/132403#issuecomment-3667073661 as an alternative. WDYT?

aryasoni98 committed:
Fix imageMaximumGCAge download/remove loop with CreateContainerConfigError

Add PodGetter to ImageGCManager so images referenced by non-terminal pods
(desired state) are considered in use. This fixes the case where the image
was pulled but container creation failed (CreateContainerConfigError) - no
container exists in the runtime, so the image was incorrectly GC'd and
re-pulled repeatedly.

Fixes #132403
